src/t8_geometry/t8_geometry_implementations/t8_geometry_occ.cxx

```diff
@@ -0,0 +1,745 @@
+      surface->D0 (interpolated_surface_params[0],
+                   interpolated_surface_params[1], pnt);
+
+      /* Compute the displacement between surface and interpolated coords, scale them with the appropriate ref_coord 
+       * and add them to the out_coords. */
+      if (i_faces % 2 == 0) {
+        out_coords[0] 
+         += (pnt.X () - interpolated_face_coords[0]) 
+          * (1 - ref_coords[i_faces / 2]);
+        out_coords[1] 
+         += (pnt.Y () - interpolated_face_coords[1]) 
+          * (1 - ref_coords[i_faces / 2]);
+        out_coords[2] 
+         += (pnt.Z () - interpolated_face_coords[2]) 
+          * (1 - ref_coords[i_faces / 2]);
+      }
+      else {
+        out_coords[0] 
+         += (pnt.X () - interpolated_face_coords[0]) 
+          * ref_coords[i_faces / 2];
+        out_coords[1] 
+         += (pnt.Y () - interpolated_face_coords[1]) 
+          * ref_coords[i_faces / 2];
+        out_coords[2] 
+         += (pnt.Z () - interpolated_face_coords[2]) 
+          * ref_coords[i_faces / 2];
+      }
+      /* *INDENT-ON* */
+    }
+  }
+}
+
+void
+t8_geometry_occ::t8_geom_evalute_jacobian (t8_cmesh_t cmesh,
+                                           t8_gloidx_t gtreeid,
+                                           const double
+                                           *ref_coords,
+                                           double *jacobian_out) const
+{
+  double              h = 1e-9;
+  double              in1[3], in2[3];
+  double              out1[3], out2[3];
+  for (int dim = 0; dim < 3; ++dim) {
+    in1[0] = ref_coords[dim];
+    in1[1] = ref_coords[dim];
+    in1[2] = ref_coords[dim];
+    in2[0] = ref_coords[dim];
+    in2[1] = ref_coords[dim];
+    in2[2] = ref_coords[dim];
+
+    if (ref_coords[dim] < h) {
+      in2[dim] += ref_coords[dim] + h;
+    }
+    else if (ref_coords[dim] > 1 - h) {
+      in1[dim] -= h;
+    }
+    else {
+      in1[dim] -= 0.5 * h;
+      in2[dim] += 0.5 * h;
+    }
+    t8_geometry_occ::t8_geom_evaluate (cmesh, gtreeid, in1, out1);
+    t8_geometry_occ::t8_geom_evaluate (cmesh, gtreeid, in2, out2);
+    for (int dim2 = 0; dim2 < 3; ++dim2) {
+      jacobian_out[dim * 3 + dim2] = out1[dim2] - out2[dim2];
+    }
+  }
+}
+
+/* Our indent skript has huge problems with c++ */
+/* *INDENT-OFF* */
+inline void
+t8_geometry_occ::t8_geom_load_tree_data (t8_cmesh_t cmesh,
+                                         t8_gloidx_t gtreeid)
+{
+  t8_geometry_w_vertices::t8_geom_load_tree_data(cmesh, gtreeid);
+  edges = (const int *) t8_cmesh_get_attribute (cmesh, t8_get_package_id (),
+                                                T8_CMESH_OCC_EDGE_ATTRIBUTE_KEY,
+                                                gtreeid);
+  faces = (const int *) t8_cmesh_get_attribute (cmesh, t8_get_package_id (),
+                                                T8_CMESH_OCC_FACE_ATTRIBUTE_KEY,
+                                                gtreeid);
```

Comment:
Should assert that edges and faces are not NULL (attribute not found).

Reply:
How do I check that? `T8_ASSERT(faces != NULL);` seems wrong to me, because faces is a pointer

Reply:
Yes that is exactly the correct assertion.
The return value of t8_cmesh_get_attribute is NULL if anything is wrong, so faces equals NULL in that case.

Reply:
Okay thanks

Reply:
resolved in 3d55c24